File: runinator-command-center-flutter/lib/core/domain/models/replica.dart
// port of core/domain/models/replica.ts.

import '../json.dart';

enum ReplicaKind {
  worker('worker'),
  waker('waker'),
  webservice('webservice'),
  postgres('postgres'),
  archiver('archiver');

  const ReplicaKind(this.wire);

  final String wire;

  static ReplicaKind fromJson(String value) => ReplicaKind.values.firstWhere(
        (kind) => kind.wire == value,
        orElse: () => throw ArgumentError('unknown ReplicaKind: $value'),
      );

  String toJson() => wire;
}

enum ReplicaStatus {
  live('live'),
  stale('stale'),
  offline('offline');

  const ReplicaStatus(this.wire);

  final String wire;

  static ReplicaStatus fromJson(String value) => ReplicaStatus.values.firstWhere(
        (status) => status.wire == value,
        orElse: () => throw ArgumentError('unknown ReplicaStatus: $value'),
      );

  String toJson() => wire;
}

class ReplicaRecord {
  const ReplicaRecord({
    required this.replicaId,
    required this.replicaType,
    required this.instanceId,
    required this.runtimeId,
    required this.status,
    this.displayName,
    this.host,
    this.port,
    this.basePath,
    this.observedIp,
    this.version,
    required this.attributes,
    required this.firstSeenAt,
    required this.lastHeartbeatAt,
    required this.lastSeenAt,
    this.offlineAt,
  });

  factory ReplicaRecord.fromJson(Map<String, Object?> json) => ReplicaRecord(
        replicaId: json['replica_id'] as String,
        replicaType: ReplicaKind.fromJson(json['replica_type'] as String),
        instanceId: json['instance_id'] as String,
        runtimeId: json['runtime_id'] as String,
        status: ReplicaStatus.fromJson(json['status'] as String),
        displayName: json['display_name'] as String?,
        host: json['host'] as String?,
        port: (json['port'] as num?)?.toInt(),
        basePath: json['base_path'] as String?,
        observedIp: json['observed_ip'] as String?,
        version: json['version'] as String?,
        attributes: asJsonObject(json['attributes']),
        firstSeenAt: json['first_seen_at'] as String,
        lastHeartbeatAt: json['last_heartbeat_at'] as String,
        lastSeenAt: json['last_seen_at'] as String,
        offlineAt: json['offline_at'] as String?,
      );

  final String replicaId;
  final ReplicaKind replicaType;
  final String instanceId;
  final String runtimeId;
  final ReplicaStatus status;
  final String? displayName;
  final String? host;
  final int? port;
  final String? basePath;
  final String? observedIp;
  final String? version;
  final JsonObject attributes;
  final String firstSeenAt;
  final String lastHeartbeatAt;
  final String lastSeenAt;
  final String? offlineAt;

  Map<String, Object?> toJson() => {
        'replica_id': replicaId,
        'replica_type': replicaType.toJson(),
        'instance_id': instanceId,
        'runtime_id': runtimeId,
        'status': status.toJson(),
        'display_name': displayName,
        'host': host,
        'port': port,
        'base_path': basePath,
        'observed_ip': observedIp,
        'version': version,
        'attributes': attributes,
        'first_seen_at': firstSeenAt,
        'last_heartbeat_at': lastHeartbeatAt,
        'last_seen_at': lastSeenAt,
        'offline_at': offlineAt,
      };
}

class ReplicaCounts {
  const ReplicaCounts({required this.workers, required this.wakers, required this.webservices});

  factory ReplicaCounts.fromJson(Map<String, Object?> json) => ReplicaCounts(
        workers: (json['workers'] as num).toInt(),
        wakers: (json['wakers'] as num).toInt(),
        webservices: (json['webservices'] as num).toInt(),
      );

  final int workers;
  final int wakers;
  final int webservices;

  Map<String, Object?> toJson() => {'workers': workers, 'wakers': wakers, 'webservices': webservices};
}

class ReplicaListResponse {
  const ReplicaListResponse({required this.counts, required this.replicas});

  factory ReplicaListResponse.fromJson(Map<String, Object?> json) => ReplicaListResponse(
        counts: ReplicaCounts.fromJson(json['counts'] as Map<String, Object?>),
        replicas: (json['replicas'] as List)
            .map((r) => ReplicaRecord.fromJson(r as Map<String, Object?>))
            .toList(),
      );

  final ReplicaCounts counts;
  final List<ReplicaRecord> replicas;

  Map<String, Object?> toJson() => {
        'counts': counts.toJson(),
        'replicas': replicas.map((r) => r.toJson()).toList(),
      };
}

File: runinator-command-center/src/core/domain/models/replica.ts
import type { JsonRecord } from "../json";

export type ReplicaKind =
  | "worker"
  | "waker"
  | "webservice"
  | "background"
  | "postgres"
  | "archiver";

export type ReplicaStatus = "live" | "stale" | "offline";

export type AgentConnectionState =
  | "stopped"
  | "registering"
  | "connecting"
  | "connected"
  | "draining"
  | "reconnecting"
  | "reenrollment_required";

export interface AgentStatusReport {
  connection_state: AgentConnectionState;
  reconnect_retry_seconds?: number | null;
  broker_mode: string;
  broker_endpoint: string;
  in_flight: number;
  succeeded: number;
  failed: number;
  timed_out: number;
  canceled: number;
  last_error?: string | null;
  last_error_at?: string | null;
  outbox_depth: number;
  agent_version?: string | null;
  config_hash: string;
  provider_count: number;
  labels: Record<string, string>;
  uptime_seconds: number;
  heartbeat_seq: number;
  clock_skew_ms: number;
  stale_after_seconds?: number | null;
}

export interface ReplicaRecord {
  replica_id: string;
  replica_type: ReplicaKind;
  instance_id: string;
  runtime_id: string;
  status: ReplicaStatus;
  display_name?: string | null;
  host?: string | null;
  port?: number | null;
  base_path?: string | null;
  observed_ip?: string | null;
  version?: string | null;
  attributes: JsonRecord;
  first_seen_at: string;
  last_heartbeat_at: string;
  last_seen_at: string;
  offline_at?: string | null;
}

/// one provider a replica advertises. the metadata is the provider's own, so it is typed loosely
/// here and read only for its name and action count.
export interface ReplicaProviderRegistration {
  replica_id: string;
  provider_name: string;
  provider: {
    name: string;
    actions: { function_name: string; description?: string | null }[];
    metadata: { credential_scopes: string[] };
  };
  created_at?: string;
  updated_at?: string;
}

export interface ReplicaCounts {
  workers: number;
  wakers: number;
  webservices: number;
  background: number;
}

export interface ReplicaListResponse {
  counts: ReplicaCounts;
  replicas: ReplicaRecord[];
}

export type AgentDirectiveState =
  | "pending"
  | "published"
  | "accepted"
  | "completed"
  | "failed"
  | "unsupported"
  | "expired";

export type AgentDirectiveKind =
  | { type: "diagnostics" }
  | { type: "tail_logs"; lines: number }
  | { type: "list_sandbox"; path: string }
  | { type: "fetch_file"; path: string; max_bytes: number }
  | { type: "drain" }
  | { type: "undrain" }
  | { type: "restart" };

export interface AgentDirectiveRecord {
  directive_id: string;
  replica_id: string;
  kind: AgentDirectiveKind | JsonRecord;
  state: AgentDirectiveState;
  issued_at: string;
  expires_at: string;
  published_at?: string | null;
  completed_at?: string | null;
  payload: unknown;
  message?: string | null;
  attempts: number;
}
